### PR TITLE
fix: prevent DatePicker from some unnecessary rerenders

### LINKED ISSUE #2546: DatePicker: Re Rendering the component cause wrong behavior of inputs
## Detailed description

When `DatePicker` has `defaultValues` - `values` and a `ranged` mode it has a bug.
After re-rendering the component in the first input field it will have `defaultValue[0]` then `rangeSeparator` (from flatpickr) and then `defaultValue[1]`. Here is the screenshot. 

image

Probably after re-rendering the component react DatePicker did not parse and split the data and then put it to the second input. To my mind problem should be somewhere here. Sadly, I did not found part of code which is responsible for that.

carbon-components: 9.42.3
carbon-components-react: 6.45.2

You could test it at codesandbox - [codesandbox](https://q4ny4275w6.codesandbox.io/)
## Additional information

Before rerednering:
before rerednering

After rerednering:
after rerednering

It is kind of important component for our team. So, it would be nice if you can fix it asap, please. Or, at least point me out what to check/debug in your code.
Thank you!

### PR DESCRIPTION
**Pull request by Steven-Gassert**
Closes #
I originally set out to close https://github.com/carbon-design-system/carbon/issues/2546

What I noticed is that between the way that `DatePicker` consumes the selected date, and the way that Flatpickr parses the selected date, this error occurs during the `UNSAFE_componentWillUpdate` hook [here](https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/components/DatePicker/DatePicker.js#L312) I was not able to determine exactly where this error was happening but did notice that the error yasotnik described seeing was most likely due to the fact that in [this hook](https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/components/DatePicker/DatePicker.js#L312) multiple date arrays would not be compared properly, thus causing unnecessary re-renders.

**TL;DR** I don't believe this would solve the issue of parsing ranged dates on all 'fully controlled' inputs, but should solve the use case where yasotnik originally saw this error (when providing default values) and should prevent some unnecessary re-renders in general

#### Changelog

**Changed**

- the way `DatePicker` compares props to decide when to update

#### Testing / Reviewing

- Note that when comparing `nextProps.value` and `this.props.value` [here](https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/components/DatePicker/DatePicker.js#L310) will return true (as in that they are not the same value) when `value` is an array of dates. This would not happen when using lodash [isEqual](https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/components/DatePicker/DatePicker.js#L310)

### CODE REVIEW

**Review by asudoh — COMMENTED** [review]
Hi 👋 thank you for jumping in! React application should change the whole object/array tree (instead of changing sub-property) when changing a value in there, and thus deep-comparison doesn't make sense for change detection in general.

**Comment by joshblack** [comment]
Hey @Steven-Gassert! 👋 Just wanted to follow-up and say thanks again for taking the time to submit this PR 🙏 

I'm going through and cleaning up our PR backlog and came across this without any new activity for about a week. Just to keep things up-to-date, I am going to close this for now but feel free to re-open if you have any response to the comments above!
